Bash to python

rustprooflabs opened the pull request:
Still to do:

* :x: <del>Install newer Python (3.9) in Dockerfile</del> ([see comment](https://github.com/rustprooflabs/pgosm-flex/pull/147#issuecomment-890540263))

Done:

* :heavy_check_mark: Prepare db
* :heavy_check_mark: run osm2pgsql
* :heavy_check_mark: post-processing steps
* :heavy_check_mark: Handle `PGOSM_SKIP_NESTED_POLYGON` env var -- Also `--skip-nested` flag
* :heavy_check_mark: Add Makefile, easier local dev/testing  - Thank you @jacopofar
* :heavy_check_mark: Add requirements.txt
* :heavy_check_mark: run pg_dump
* :heavy_check_mark: Handle `-latest` files pre/post import
* :heavy_check_mark: Handle `PGOSM_DATA_SCHEMA_ONLY` env var -- Toggles Sqitch deployment & QGIS styles
* :heavy_check_mark: Handle Env var - skip nested polygon
* :heavy_check_mark: Enable `PGOSM_DATA_SCHEMA_NAME` via `--schema-name`  - impacts `pg_dump`
* :heavy_check_mark: Enable `PGOSM_SRID` via `--srid` -- Want it documented with `--help` command
* :heavy_check_mark: Enable `PGOSM_LANGUAGE` via `--language` -- Want it documented with `--help` command
* :heavy_check_mark: Update documentation to reflect new style of options

Removed option

* `ALWAYS_DOWNLOAD` was remove.  Users can `rm <your_files>` easy enough in that (hopefully) rare occurance.

Makes new features easier to implement

*  #131
* #148

jacopofar commented:
Nice! I am looking forward to it, I work with python and maybe I can help?

rustprooflabs commented:
> Nice! I am looking forward to it, I work with python and maybe I can help?

@jacopofar That would be fantastic, thank you!  If you can fork and PR back into this branch that would be great.  The code is plenty rough right now, take your pick of adding missing features from the bash process and/or making other improvements you see fit to the code style/structure.  I've tried to leave some breadcrumbs in the code to where some loose ends are.

Let me know if there's anything I can do to help you move forward!

jacopofar commented on docker/README.md:
```diff
@@ -0,0 +1,31 @@
+# Run Unit tests
+
+Requires Python packages installed by `Dockerfile`. Look for `
```
Maybe a Makefile could include all these steps? the virtualenv could be created on a folder inside the project so everything remains self-contained

rustprooflabs replied:
Great idea.

jacopofar commented on docker/pgosm_flex.py:
```diff
@@ -0,0 +1,318 @@
+    log_file = get_log_path(region, subregion, paths)
+    print(f'Logging output to {log_file}')
+
+    # Docker image has Python 3.7, this doesn't work for some reason.
```
the encoding parameter was added in Python 3.9, I think in this dockerized environment is utf-8 anyways

jacopofar commented:
I did a small PR #149 with some minor changes. I got stuck when it tries to create the DB, this is the command and the output from inside the container:


```
root@36d8e28dc981:/app# python3 docker/pgosm_flex.py --layerset run-all --ram 8 --region north-america/us --subregion district-of-columbia
Running PgOSM Flex!
Logging output to /app/output/north-america-us-district-of-columbia.log
PgOSM Flex starting...
PBF file not found locally. Download required
Downloading PBF and MD5 files...
Downloading PBF data to /app/output/district-of-columbia-latest.osm.pbf
osm2pgsql-tuner URL w/ parameters: https://osm2pgsql-tuner.com/api/v1?system_ram_gb=8&osm_pbf_gb=0.016266328282654285&append=False&pbf_filename=/app/output/district-of-columbia-latest.osm.pbf&pgosm_layer_set=run-all
Status code: 200
osm2pgsql -U postgres -d pgosm  --output=flex --style=./run-all.lua  /app/output/app/output/district-of-columbia-latest.osm.pbf.osm.pbf
FIXME: Not fully functional!
Checking for Postgres service to be available
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Postgres still has not started. Exiting.
```
I saw that FIXME but didn't understand what was supposed to happen so I could not continue and try to implement or avoid it.

A few things I could try to work on:
* add a makefile for the project that also includes creating and configuring a local virtualenv if running outside Docker, so one can run tests from scratch, and if the virtualenv is local it can be safely deleted afterwards
* adding a requirements.txt with the Python requirements versions pinned
* try to move to python 3.9 and installing pip with the other packages (should be the python3-pip package on this image)

rustprooflabs commented:
> I did a small PR #149 with some minor changes. I got stuck when it tries to create the DB, this is the command and the output from inside the container:
> 
> ```
> root@36d8e28dc981:/app# python3 docker/pgosm_flex.py --layerset run-all --ram 8 --region north-america/us --subregion district-of-columbia
> Running PgOSM Flex!
> Logging output to /app/output/north-america-us-district-of-columbia.log
> PgOSM Flex starting...
> PBF file not found locally. Download required
> Downloading PBF and MD5 files...
> Downloading PBF data to /app/output/district-of-columbia-latest.osm.pbf
> osm2pgsql-tuner URL w/ parameters: https://osm2pgsql-tuner.com/api/v1?system_ram_gb=8&osm_pbf_gb=0.016266328282654285&append=False&pbf_filename=/app/output/district-of-columbia-latest.osm.pbf&pgosm_layer_set=run-all
> Status code: 200
> osm2pgsql -U postgres -d pgosm  --output=flex --style=./run-all.lua  /app/output/app/output/district-of-columbia-latest.osm.pbf.osm.pbf
> FIXME: Not fully functional!
> Checking for Postgres service to be available
> Waiting...
> Waiting...
> Waiting...
> Waiting...
> Waiting...
> Waiting...
> Waiting...
> Postgres still has not started. Exiting.
> ```
> 
> I saw that FIXME but didn't understand what was supposed to happen so I could not continue and try to implement or avoid it.
> 

#149 is merged, thank you.  Ah, the `pg_isready` command is expecting Postgres running locally as it would be in Docker.  That should probably be updated to accept connection string details, e.g. `pg_isready -h your_pg_host -U your_pg_user`.

> A few things I could try to work on:
> 
> * add a makefile for the project that also includes creating and configuring a local virtualenv if running outside Docker, so one can run tests from scratch, and if the virtualenv is local it can be safely deleted afterwards
> * adding a requirements.txt with the Python requirements versions pinned
> * try to move to python 3.9 and installing pip with the other packages (should be the python3-pip package on this image)

These all sound great.  I'll keep working on finishing the program flow to kick off osm2pgsql and export the data.

rustprooflabs commented:
RE local testing w/out docker - There are (at least) 3 assumptions that are made

* Postgres is running locally
* your current user has createdb in local Pg
* osm2pgsql >=1.5.0 is installed

Testing locally should be done with caution, these scripts are written (at least by me) with the expectation of being contained within Docker.  In general, testing via docker will be encouraged.  I've stubbed my own toe the last few days against other minor differences on "my machine" vs Docker so want to avoid too much of that.

My workflow for testing this in Docker is 4 commands.  Probably need to work these into some better docs.

```bash
docker build -t rustprooflabs/pgosm-flex .
docker stop pgosm && docker rm pgosm
docker run --name pgosm -d     -v ~/pgosm-data:/app/output     -v /etc/localtime:/etc/localtime:ro     -e POSTGRES_PASSWORD=mysecretpassword     -p 5433:5432 -d rustprooflabs/pgosm-flex
docker exec -it     -e POSTGRES_PASSWORD=mysecretpassword -e POSTGRES_USER=postgres     pgosm python3 docker/pgosm_flex.py     --layerset=run-all --ram=8     --region=north-america/us     --subregion=district-of-columbia
```

I pushed another small commit that cleans out some cruft and mocks the next functions needed.

rustprooflabs commented:
@jacopofar RE Python 3.9 -- Let's hold off and stick with Python 3.7 unless there is a compelling reason to upgrade.  The closer we stick to the base Postgres/PostGIS image the better.  I'm open to reasons for upgrading, but my initial comment/issue was obviously an easy fix to get working in 3.7.

I reworked the logging a bit and it's working as expected in both Docker and locally now. My current local-test Python command is this including my local-specific path.

```python
python3 pgosm_flex.py  \
 --layerset=run-all --ram=8 \
 --region=north-america/us \
  --subregion=district-of-columbia \
  --basepath=/home/ryanlambert/git/pgosm-flex \
  --debug
```

rustprooflabs commented:
Latest changes make the critical path functional to run osm2pgsql resulting in data being loaded.  Still not running post processing SQL, loading additional structures, running pg_dump, or full file archive handling. 

```bash
docker pull rustprooflabs/pgosm-flex:dev-python
```

rustprooflabs commented:
Added an `all` target as default to Makefile.  Running `make` now cleans & builds.

Added requirements.txt, used by Dockerfile.

Update osm2pgsql build to run in parallel based on `nproc` output.  Cuts nearly 2 minutes off full rebuild on my laptop.

rustprooflabs commented:
Not all of the options have been re-enabled yet, but the core of the rewrite is functionally complete through pg_dump.

Updated docker image:

```
docker pull rustprooflabs/pgosm-flex:dev-python
```

rustprooflabs commented:
@jacopofar The Python process should be feature complete at this point.  I logged a new bug (#153) for the one step that works in normal operation but not with the Makefile's modified user RE file permissions.

The python code is messy but functional.  My next steps are to go through the various readme's and update the various instructions for the new command format.

Docker image on hub is updated

```
docker pull rustprooflabs/pgosm-flex:dev-python
```
